packages/ember-routing/tests/system/controller_for_test.js

```diff
@@ -1,112 +1,59 @@
-    namespace = Namespace.create();
-    appInstance = buildInstance(namespace);
-    appInstance.register('controller:app', Controller.extend());
-    appController = appInstance.lookup('controller:app');
-  },
-  teardown() {
-    run(() => {
-      appInstance.destroy();
-      namespace.destroy();
+    return this.visit('/').then(() => {
+      let appInstance = this.applicationInstance;
+      let appController = appInstance.lookup('controller:app');
+      let controller = controllerFor(appInstance, 'app');
+      assert.equal(appController, controller, 'should find app controller');
     });
   }
 });
 
-QUnit.test('controllerFor should lookup for registered controllers', function() {
-  let controller = controllerFor(appInstance, 'app');
-
-  equal(appController, controller, 'should find app controller');
-});
-
-QUnit.module('Ember.generateController', {
-  setup() {
-    namespace = Namespace.create();
-    appInstance = buildInstance(namespace);
-  },
-  teardown() {
-    run(() => {
-      appInstance.destroy();
-      namespace.destroy();
+moduleFor('Ember.generateController', class extends ApplicationTestCase {
+  ['@test generateController should return Ember.Controller'](assert) {
+    return this.visit('/').then(() => {
+      let controller = generateController(this.applicationInstance, 'home');
+      assert.ok(controller instanceof Controller, 'should return controller');
     });
   }
-});
 
-QUnit.test('generateController should return Ember.Controller', function() {
-  let controller = generateController(appInstance, 'home');
+  ['@test generateController should return App.Controller if provided'](assert) {
+    let MainController = Controller.extend();
+    this.add('controller:basic', MainController);
 
-  ok(controller instanceof Controller, 'should return controller');
-});
-
-
-QUnit.test('generateController should return App.Controller if provided', function() {
-  let controller;
-  namespace.Controller = Controller.extend();
-
-  controller = generateController(appInstance, 'home');
-
-  ok(controller instanceof namespace.Controller, 'should return controller');
-});
-
-QUnit.test('generateController should return controller:basic if provided', function() {
-  let controller;
-
-  let BasicController = Controller.extend();
-  appInstance.register('controller:basic', BasicController);
-
-  controller = generateController(appInstance, 'home');
+    return this.visit('/').then(() => {
+      let controller = generateController(this.applicationInstance, 'home');
+      assert.ok(controller instanceof MainController, 'should return controller');
+    });
+  }
 
-  if (EMBER_NO_DOUBLE_EXTEND) {
-    ok(controller instanceof BasicController, 'should return base class of controller');
-  } else {
-    let doubleExtendedFactory;
-    ignoreDeprecation(() => {
-      doubleExtendedFactory = appInstance._lookupFactory('controller:basic');
+  ['@test generateController should return controller:basic if provided'](assert) {
```

Comment:
This test was poorly named. It was testing the case of a registered factory. The test above it tests the resolved case.

As such, this test should not use `this.add('controller:basic'` but instead should register the factory:

* `this.application.register('controller:basic'`
* Even more like the old test would be `this.applicationInstance.register('controller:basic'` after the `visit` call.

Finally the test name should be `@test generateController should return controller:basic if registered` (note "registered" vs. "resolved" in the above test).